### PR TITLE
Add scaled_dot_product_attention op

### PR DESCRIPTION
**Pull request by cglagovichTT**
#7461

### CODE REVIEW

**Review by TT-BrianLiu — COMMENTED** [review]
General comments since there's a lot of code:
- Please switch SPDX license to 2024 for new files
- Clean up all kernels
- Add a test for program_cache in unit tests.

**Comment by cglagovichTT** [comment]
Post commit passed https://github.com/tenstorrent/tt-metal/actions/runs/8853951289

I cleaned up the kernels and updated copyright @TT-BrianLiu

File: models/experimental/llama2_70b/tests/unit_tests/test_flash_sdpa.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+        mine = tt_fa2_noconvert(device, tt_q, tt_k, tt_v, tt_attn_mask, program_config)
+        out_pass, out_pcc = comp_pcc(gt, mine, 0.99)
+        if i == 0:
+            first_pcc = out_pcc
+        else:
+            # assert out_pcc == first_pcc, "ND PCC found"
+            pass
+
+        print(f"python vs pytorch: {out_pcc}")
+        # assert out_pass
```
Comment (TT-BrianLiu) [inline models/experimental/llama2_70b/tests/unit_tests/test_flash_sdpa.py:395]: Get rid of prints and replace with logger.debug for cleanliness.
For testing program cache, you should:
- Assert on number of expected caches (`device.num_program_cache_entries()`)
- Try to shift the memory of your inputs when you loop so you actually test the functionality of the callback

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_scaled_dot_product_attention.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    ),
+)
+def test_sdpa_tt_with_program_cache(device, b, nh, nkv, s, d, q_chunk_size, k_chunk_size, dtype, use_program_cache):
+    if (s % q_chunk_size != 0) or (s % k_chunk_size != 0):
+        pytest.skip("s must be divisible by q_chunk_size and k_chunk_size")
+    if nh == 8 and q_chunk_size == 128 and k_chunk_size == 128:
+        pytest.skip("Can cause OOM if profiling is enabled.")
+
+    for _ in range(2):
+        run_test_sdpa_tt(device, b, nh, nkv, s, d, q_chunk_size, k_chunk_size, dtype)
```
Comment (TT-BrianLiu) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_scaled_dot_product_attention.py:102]: Same comment about program cache applies here.

File: tt_eager/tt_dnn/op_library/sdpa/kernels/compute/sdpa.cpp [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+    uint32_t output_num_tiles = M * N;
+    uint32_t out_subblock_num_tiles = subblock_h * subblock_w;
+    uint32_t in0_index_offset = 0;
+    uint32_t in1_index_offset = 0;
+
+    for (uint32_t in0_subblock = 0; in0_subblock < in0_num_subblocks; ++in0_subblock) {
+        for (uint32_t in1_subblock = 0; in1_subblock < in1_num_subblocks; ++in1_subblock) {
+            tile_regs_acquire();
+
+            int dst_index = 0;
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/sdpa/kernels/compute/sdpa.cpp:294]: uint32_t

File: tt_eager/tt_dnn/op_library/sdpa/kernels/compute/sdpa.cpp
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/sdpa/kernels/compute/sdpa.cpp:1]: Is there a reason some functions are templated and some are passed in as args?

File: tt_eager/tt_dnn/op_library/sdpa/kernels/dataflow/reader_interleaved.cpp [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    constexpr uint32_t q_tile_bytes = get_tile_size(cb_q_in);
+    constexpr DataFormat q_data_format = get_dataformat(cb_q_in);
+    constexpr uint32_t k_tile_bytes = get_tile_size(cb_k_in);
+    constexpr DataFormat k_data_format = get_dataformat(cb_k_in);
+    constexpr uint32_t v_tile_bytes = get_tile_size(cb_v_in);
+    constexpr DataFormat v_data_format = get_dataformat(cb_v_in);
+    constexpr uint32_t mask_tile_bytes = get_tile_size(cb_mask_in);
+    constexpr DataFormat mask_data_format = get_dataformat(cb_mask_in);
+
+    const uint32_t barrier_threshold = get_barrier_read_threshold(q_tile_bytes, num_cores);
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/sdpa/kernels/dataflow/reader_interleaved.cpp:63]: Is `num_cores` dynamic? If not, we can make this compile-time or `constexpr`

File: tt_eager/tt_dnn/op_library/sdpa/kernels/dataflow/writer_interleaved.cpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    constexpr uint32_t tile_bytes = get_tile_size(cb_out);
+    constexpr DataFormat data_format = get_dataformat(cb_out);
+
+    const InterleavedAddrGenFast<is_dram> out_writer = {
+        .bank_base_address = out_addr,
+        .page_size = tile_bytes,
+        .data_format = data_format
+    };
+
+    const uint32_t barrier_threshold = get_barrier_read_threshold(tile_bytes, num_cores);
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/sdpa/kernels/dataflow/writer_interleaved.cpp:54]: same comment as reader

File: tt_eager/tt_dnn/op_library/sdpa/kernels/dataflow/writer_interleaved.cpp [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    uint32_t out_tile_id = 0;
+
+    for (uint32_t nb = local_batch_start; nb < local_batch_end; ++nb) {
+        const uint32_t q_batch_offset = nb * NQH * St * DHt;
+        for (uint32_t nq = local_nh_start; nq < local_nh_end; ++nq) {
+            for (uint32_t q_iter = 0; q_iter < q_chunks_per_core; ++q_iter) {
+                uint32_t q_chunk;
+                #if defined BALANCED_Q_PARALLEL
+                if (q_iter < q_chunks_per_core / 2) {
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/sdpa/kernels/dataflow/writer_interleaved.cpp:71]: precompute `q_chunks_per_core / 2`

File: tt_eager/tt_dnn/op_library/sdpa/multi_core/sdpa_op_multi_core.cpp
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/sdpa/multi_core/sdpa_op_multi_core.cpp:1]: We should switch all `TT_ASSERT` to `TT_FATAL`

**Review by TT-BrianLiu — APPROVED** [review]
Minor comments. Please address before merge

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_scaled_dot_product_attention.py (hunk 2 above)
Comment (cglagovichTT) [reply tests/tt_eager/python_api_testing/unit_testing/misc/test_scaled_dot_product_attention.py:102]: In this test I run the test twice with program caching enabled, so the op should run on 2 different sets of inputs at different addresses. that should test it right

File: tt_eager/tt_dnn/op_library/sdpa/kernels/compute/sdpa.cpp
Comment (cglagovichTT) [reply tt_eager/tt_dnn/op_library/sdpa/kernels/compute/sdpa.cpp:1]: I figured that functions with constexpr inputs might as well be templates to give the compiler a chance to optimize based on input sizes. This gave me a very small gain for a few functions

**Pull request merged**